File: solutions/elixir/allergies/1/lib/allergies.ex
use Bitwise

defmodule Allergies do
  defp allergens do
    ~w[eggs peanuts shellfish strawberries tomatoes chocolate pollen cats] 
    |> Enum.with_index
    |> Map.new(fn {e, idx} -> {Integer.pow(2, idx), e} end)
  end
  
  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags) do
    allergens()
    |> Map.filter(fn {bit, _} -> (bit &&& flags) != 0 end)
    |> Map.values
    |> MapSet.new
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item) do
    list(flags) |> MapSet.member?(item)
  end
end

File: solutions/scala/allergies/1/src/main/scala/Allergies.scala
enum Allergen(val score: Int) {
  case Eggs extends Allergen(1)
  case Peanuts extends Allergen(2)
  case Shellfish extends Allergen(4)
  case Strawberries extends Allergen(8)
  case Tomatoes extends Allergen(16)
  case Chocolate extends Allergen(32)
  case Pollen extends Allergen(64)
  case Cats extends Allergen(128)
}

object Allergies {
  def allergicTo(allergen: Allergen, score: Int): Boolean =
    (score & allergen.score) != 0

  def list(score: Int): List[Allergen] =
    Allergen.values.filter(allergicTo(_, score)).toList
}
